leakpro/attacks/mia_attacks/yoqo.py

```diff
@@ -0,0 +1,340 @@
+        loss = torch.cat([shadow_model.model_obj(x0 + dx) for shadow_model in self.shadow_models], dim = 1)
+        loss = self.loss(loss,target_output)
+        loss = (loss * weights).sum()
+        norm =  self.alpha * self.mse(dx, torch.zeros(dx.shape, device = dx.device))
+        return (loss + norm)
+
+    def _optimize_xprime(self: Self, x0: Tensor, target_output: Tensor, weights: Tensor) -> Tensor:
+        dx = torch.zeros(x0.shape, device = x0.device)
+        dx.requires_grad = True
+
+        optim = torch.optim.SGD([dx], lr = self.lr_xprime_optimization)
+        for _ in range(self.max_iterations):
+            loss = self._optimization_objective(x0, dx, target_output, weights)
+            if loss < self.stop_criterion:
+                break
+            loss.backward()
+            optim.step()
+            optim.zero_grad()
+
+        return (x0 + dx)
+
+    def run_attack(self:Self) -> CombinedMetricResult:
+        """Runs the attack on the target model and dataset and assess privacy risks or data leakage.
+
+        This method evaluates how the target model's output (logits) for a specific dataset
+        compares to the output of shadow models to determine if the dataset was part of the
+        model's training data or not.
+
+        Returns
+        -------
+        Result(s) of the metric. An object containing the metric results, including predictions,
+        true labels, and signal values.
+
+        """
+
+        data_loader = self.handler.get_dataloader(self.audit_data_indices, batch_size=1)
+
+        predictions = []
+        signal_values = []
+
+        device_name = torch.device("cuda" if torch.cuda.is_available() else "mps" if torch.mps.is_available() else "cpu")
+
+        for model in self.shadow_models:
+            model.model_obj.eval()
+            model.model_obj.to(device_name)
+
+        for i, (data, labels) in tqdm(enumerate(data_loader), desc="Optimizing queries", leave=False):
+            xprime = self._optimize_xprime(data.to(device_name),
+                                           self.target_output[i:(i + 1)].to(device_name),
+                                           self.weights[i:(i + 1), :].to(device_name))
+
+            lprime = self.target_model.get_logits(xprime)
+            lprime = np.argmax(lprime, axis = 1)
+
+            batch_predictions = np.equal(lprime,labels)
+            predictions.extend(batch_predictions)
+
+        predictions = np.array(predictions).reshape(1,-1)
+        signal_values = predictions.copy().reshape(-1,1)
+
+        for model in self.shadow_models:
+            model.model_obj.train()
+
+        # Prepare true labels array, marking 1 for training data and 0 for non-training data
+        true_labels = np.concatenate(
+            [np.ones(len(self.in_members)), np.zeros(len(self.out_members))]
+        )
+
+        logger.info(f"Accuracy: {np.sum(predictions == true_labels)/predictions.size}")
+
+        predictions = np.concatenate(
```

Comment:
why padding zeros and ones?

Reply:
This is needed for the report generation. Otherwise I think there was a crash on trying to generate the ROC curve. Will add a comment.